<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.5.0 -->
<title>ddRAD Data in AftrRAD | Genomics of Biodiversity @ IBS2019 in Quito, Ecuador</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="ddRAD Data in AftrRAD" />
<meta name="author" content="Becca Tarvin" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador." />
<meta property="og:description" content="This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador." />
<link rel="canonical" href="http://localhost:4000/old_posts/2017-08-03-morning-AftrRAD.html" />
<meta property="og:url" content="http://localhost:4000/old_posts/2017-08-03-morning-AftrRAD.html" />
<meta property="og:site_name" content="Genomics of Biodiversity @ IBS2019 in Quito, Ecuador" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2017-08-04T00:00:00-07:00" />
<script type="application/ld+json">
{"description":"This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador.","author":{"@type":"Person","name":"Becca Tarvin"},"@type":"BlogPosting","url":"http://localhost:4000/old_posts/2017-08-03-morning-AftrRAD.html","headline":"ddRAD Data in AftrRAD","dateModified":"2017-08-04T00:00:00-07:00","datePublished":"2017-08-04T00:00:00-07:00","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/old_posts/2017-08-03-morning-AftrRAD.html"},"@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <link rel="stylesheet" href="/assets/css/style.css?v=a2fbe73ec3016e9099eea62f3c507e0f936c7fba">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Genomics of Biodiversity @ IBS2019 in Quito, Ecuador</h1>
      <h2 class="project-tagline">This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador.</h2>
      
        <a href="https://github.com/rdtarvin/IBS2019_Genomics-of-Biodiversity" class="btn">View on GitHub</a>
      
      
    </section>

    <section class="main-content">
      <h2 id="workshop-quito---day-5-">WORKSHOP QUITO - DAY 5 <br /></h2>
<h3 id="ddrad-data-in-aftrrad">ddRAD data in AftrRAD</h3>

<p>AftrRAD (align from total reads) is a bioinformatic pipeline for the de novo assembly and genotyping of RADseq data. It was described by Sovic et al (2015), Molecular Ecology Resources 15:1163-1171, and you can get the program and documentation as a zip file <a href="https://u.osu.edu/sovic.1/downloads/">here</a>.</p>

<h2 id="part-0-getting-ready-for-aftrrad">PART 0 Getting ready for AftrRAD</h2>

<p>AftrRAD is a good alternative when you have single-end RADseq data and no access to multicore computing resources (i.e. you can perform these analyses efficiently on a personal laptop).</p>

<p>In order to run AftrRAD you need to unzip the file downloaded before and make sure the three dependencies (i.e. R, ACANA, and Mafft) mentioned in the manual are installed and working correctly. Also, if running AftrRAD version 5.0 or greater you can reduce run times with  parallel analyses by installing the Perl module Parallel:ForkManager.</p>

<p><strong>NOTE</strong>: All of the steps mentioned above have already been taken care of when installing the Virtual Machine for this workshop.</p>

<p><img src="https://github.com/rdtarvin/RADseq_Quito_2017/blob/master/images/basic-assembly-steps.png?raw=true" alt="" /><br /></p>

<h2 id="part-1-formatting-data-and-steps-1-2-and-3-aftrradpl">PART 1 Formatting data and steps 1, 2 and 3* (AftrRAD.pl)</h2>

<p>Download data for this lesson and include the file in a folder called Data.</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cd ~/Applications/AftrRADv5.0/Data/
wget -O RAD105_10M.txt.gz 'https://www.dropbox.com/s/tqctvehvbjk0qgt/RAD105_10M.txt.gz?dl=1'
gunzip RAD105_10M.txt.gz
ls
</code></pre></div></div>

<p>Create a text file containing the barcode and sample name information for each individual. Include the file in a folder called Barcodes.</p>

<p>Before running the first perl script make sure your working directory contains the following elements:</p>

<p><img src="https://github.com/rdtarvin/RADseq_Quito_2017/blob/master/images/AftrRAD%20working%20directory.png?raw=true" alt="" /><br /></p>

<p>Start running AftrRAD by typing perl AftrRAD.pl followed by any arguments you want to modify. Arguments and their values should be separated by a dash.</p>

<p>Arguments for this perl script are:</p>

<p><strong>re</strong>  Restriction enzyme recognition sequence. Default: TGCAGG (corresponds to SbfI). If no restriction enzyme recognition site, enter 0.</p>

<p><strong>minQual</strong> Minimum quality (Phred) score for retaining reads. Default: 20.</p>

<p><strong>minDepth</strong>  Minimum mean nonzero read depth to retain read. Default: 5.</p>

<p><strong>minIden</strong>	Minimum percent identity to consider two reads alternative alleles from same locus. Default: 90%.</p>

<p><strong>numIndels</strong>	Maximum number indels allowed between any two reads to consider them alternative alleles from the same locus. Default: 3.</p>

<p><strong>P2</strong>	Beginning of P2 adaptor sequence. Reads containing this are removed. Default: ATTAGATC.</p>

<p><strong>minParalog</strong>	Minimum number reads that must occur at a third allele in an individual to flag a locus as paralogous. Default: 5.</p>

<p><strong>Phred</strong>	Quality score methodology used in sequencing. Default: Phred33.</p>

<p><strong>dplexedData</strong>	If data are demultiplexed prior to run, set to ‘1’.</p>

<p><strong>stringLength</strong>	Reads containing strings of homopolymers of this length will be removed. Default: 15.</p>

<p><strong>DataPath</strong>	Path to directory containing fastq data files for the run. Default: Data directory in AftrRAD working directory.</p>

<p><strong>BarcodePath</strong>	Path to directory containing barcode files. Default: Barcodes directory in AftrRAD working directory.</p>

<p><strong>MaxH</strong>	Maximum proportion of samples allowed to be heterozygous at a locus. Default: 90%.</p>

<p><strong>maxProcesses</strong>	Maximum number of processors to use in a parallel run.</p>

<h2 id="part-2-steps-4-and-5-genotypepl">PART 2 Steps 4 and 5* (Genotype.pl)</h2>

<p>For the next step in AftrRAD you will need to type perl Genotype.pl followed by any arguments you want to modify. As above, arguments and their values should be separated by a dash.</p>

<p>Arguments for this perl script are:</p>

<p><strong>MinReads</strong>	Minimum coverage required at a locus in an individual to apply a binomial test and call a genotype. Default: 10.</p>

<p><strong>pvalLow</strong>	For each locus in each individual, a binomial test is applied to score the sample as heterozygous or homozygous (assumption: two alleles in a heterozygote will be sequenced in equal frequencies). Default: 0.0001.</p>

<p><strong>pvalHigh</strong>	Same as pvalLow, but allows for a different p-value threshold for loci that have relatively high total counts. Default: 0.00001.</p>

<p><strong>pvalThresh</strong> Threshold number of reads at a locus in each binomial test that determines whether pvalLow or pvalHigh is used as the critical p-value. Default: 100.</p>

<p><strong>subset</strong> Option to genotype only a subset of the individuals in dataset. Default: ‘0’, and includes all samples. If set to ‘1’, a text file must be provided with the names of the samples to include. Default name for this file is ‘SamplesForSubset.txt’, and it should be located in the main AftrRAD directory.</p>

<p><strong>subsetfile</strong> File name/path containing file with sample names to include in genotyping. Only applies if ‘1’ is set above. Default: file ‘SamplesForSubset.txt’, in the main AftrRAD directory.</p>

<p><strong>maxProcesses</strong>	Maximum number of processors to use in a parallel run.</p>

<p>This script will ask whether to eliminate specific samples based on the amount of missing data at polymorphic loci in each of the samples. Name of samples provided as “bad samples” are suggestions and to make a decision check the “MissingDataProportions.txt” and “MissingDataCounts” files present in the “Output/RunInfo” directory.</p>

<h2 id="part-3-steps-6-and-7-filtersnpspl-and-formatting-scripts-for-downstream-analyses">PART 3 Steps 6 and 7* (FilterSNPs.pl and formatting scripts for downstream analyses)</h2>

<p>Finally, you will need to type perl FilterSNPs.pl followed by any arguments you want to modify.</p>

<p>Arguments for this perl script are:</p>

<p><strong>pctScored</strong>	Percent of individuals that must be genotyped in order to retain the locus. Default: 100.</p>

<p><strong>maxSNP</strong>	Maximum location along the reads to score SNPs. Value should be chosen based on the file Output/RunInfo/SNPLocations.pdf. Default:0, which prints all SNPs.</p>

<p><strong>MinReads</strong>	Minimum coverage required at a locus in an individual to score a genotype. Default is to use the value used in the most recent run of Genotypes.pl.</p>

<p>In AftrRAD, you can find different perl scripts to prepare input files for phylogenetic and population genetic analyses. Go to the Formatting folder located in the main AftrRAD directory and type perl ScriptNameThatYouAreInterested.pl.</p>


      <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="https://github.com/rdtarvin/IBS2019_Genomics-of-Biodiversity">IBS2019_Genomics-of-Biodiversity</a> is maintained by <a href="https://github.com/rdtarvin">rdtarvin</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer>
    </section>

    
  </body>
</html>
