<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.5.0 -->
<title>2bRAD Data Pipeline | Genomics of Biodiversity @ IBS2019 in Quito, Ecuador</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="2bRAD Data Pipeline" />
<meta name="author" content="Becca Tarvin" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador." />
<meta property="og:description" content="This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador." />
<link rel="canonical" href="http://localhost:4000/old_posts/2017-08-02-afternoon-2bRAD-pyrad.html" />
<meta property="og:url" content="http://localhost:4000/old_posts/2017-08-02-afternoon-2bRAD-pyrad.html" />
<meta property="og:site_name" content="Genomics of Biodiversity @ IBS2019 in Quito, Ecuador" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2017-08-02T00:00:00-07:00" />
<script type="application/ld+json">
{"description":"This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador.","author":{"@type":"Person","name":"Becca Tarvin"},"@type":"BlogPosting","url":"http://localhost:4000/old_posts/2017-08-02-afternoon-2bRAD-pyrad.html","headline":"2bRAD Data Pipeline","dateModified":"2017-08-02T00:00:00-07:00","datePublished":"2017-08-02T00:00:00-07:00","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/old_posts/2017-08-02-afternoon-2bRAD-pyrad.html"},"@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <link rel="stylesheet" href="/assets/css/style.css?v=a2fbe73ec3016e9099eea62f3c507e0f936c7fba">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Genomics of Biodiversity @ IBS2019 in Quito, Ecuador</h1>
      <h2 class="project-tagline">This the course website for the Genomics of Biodiversity workshop at the International Biogeography Society meeting in Quito, Ecuador.</h2>
      
        <a href="https://github.com/rdtarvin/IBS2019_Genomics-of-Biodiversity" class="btn">View on GitHub</a>
      
      
    </section>

    <section class="main-content">
      <h2 id="workshop-quito---day-3">WORKSHOP QUITO - DAY 3</h2>
<h3 id="2brad-data-mixed-pipeline-from-raw-reads-to-phylogenies">2bRAD data mixed pipeline: from raw reads to phylogenies</h3>

<p>These data are part of a pilot project comparing ddRAD and 2bRAD data (<strong>do NOT distribute</strong>).<br />
There are twelve samples from three genera, with at least two individuals sampled per genus, and two technical replicates.</p>

<h3 id="download-data-for-this-lesson">Download data for this lesson</h3>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> <span class="c"># go to root of file system</span>
mkdir workshop 
<span class="nb">cd </span>workshop
<span class="c"># use ctrl+shift+v to paste within the VM</span>
wget <span class="nt">-O</span> 2bRAD.zip <span class="s1">'https://www.dropbox.com/sh/z2l0w2dq89oo55i/AAD_29lBe0MvLYLdxDB4Vm-2a?dl=1'</span>
<span class="c"># wget is a way to transfer files using a url from the command line</span>
<span class="c"># -O option makes sure the download doesn't have a wonky name like AAD_29lBe0MvLYLdxDB4Vm-2a</span>
</code></pre></div></div>

<h3 id="looking-at-your-raw-data-in-the-linux-environment">Looking at your raw data in the linux environment</h3>

<p>Your files will likely be gzipped and with the file extension <strong>.fq.gz</strong> or <strong>fastq.gz</strong>. The first thing you want to do is look at the beginning of your files while they are still gzipped.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">ls</span> <span class="c"># list all files in current directory</span>
<span class="nb">ls</span> .. <span class="c"># list all files in one directory above</span>
unzip 2bRAD.zip
<span class="nb">cd </span>2bRAD
<span class="nb">ls</span> <span class="nt">-lht</span> <span class="c"># view all files with details and in order by when they were modified, with human-readable file sizes</span>
rm 2bRAD.zip <span class="c"># IMPORTANT: removing files from the command line is permanent!!!! There is no trash</span>
zless T36R59_I93_S27_L006_R1_sub12M.fastq.gz <span class="c"># press 'q' to exit</span>
</code></pre></div></div>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head T36R59_I93_S27_L006_R1_sub12M.fastq
</code></pre></div></div>

<p>This is the fastq format, which has four lines.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>@K00179:73:HJCTJBBXX:6:1101:25905:1226 1:N:0:TGTTAG 
TNGACCAACTGTGGTGTCGCACTCACTTCGCTGCTCCTCAGGAGACAGAT 
+ 
A!AFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ 

<span class="c"># sequencer name:run ID:flowcell ID:flowcell lane:tile number in flow cell:x-coordinate of cluster :y-coordinate pair member:filtered?:control data:index sequence</span>
<span class="c"># DNA sequence</span>
<span class="c"># separator, can also sometimes (not often) hold extra information about the read</span>
<span class="c"># quality score for each base</span>
</code></pre></div></div>

<p>Ok, now lets look at the full file:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cat </span>T36R59_I93_S27_L006_R1_sub12M.fastq
</code></pre></div></div>

<p>Uh oh… let’s quit before the computer crashes… it’s too much to look at! <code class="highlighter-rouge">Ctrl+C</code><br /><br />
This is the essential computational problem with NGS data that is so hard to get over. You can’t
open a file in its entirety! Here are some alternative ways to view parts of a file at a time.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># print the first 10 lines of a file</span>
head T36R59_I93_S27_L006_R1_sub12M.fastq 

<span class="c"># print the first 20 lines of a file</span>
head <span class="nt">-20</span> T36R59_I93_S27_L006_R1_sub12M.fastq <span class="c"># '-20' is an argument for the 'head' program</span>

<span class="c"># print lines 190-200 of a file</span>
head <span class="nt">-200</span> T36R59_I93_S27_L006_R1_sub12M.fastq | tail <span class="c"># 'pipes' the first 200 lines to a program called tail, which prints the last 10 lines</span>

<span class="c"># view the file interactively</span>
less T36R59_I93_S27_L006_R1_sub12M.fastq <span class="c"># can scroll through file with cursor, page with spacebar; quit with 'q'</span>
<span class="c"># NOTE: here we can use less because the file is not in gzipped (remember that required the 'zless' command)</span>

<span class="c"># open up manual for less program</span>
man less <span class="c"># press q to quit</span>

<span class="c"># print only the first 10 lines and only the first 30 characters of each line</span>
head <span class="nt">-200</span> T36R59_I93_S27_L006_R1_sub12M.fastq | cut <span class="nt">-c</span> <span class="nt">-30</span> 

<span class="c"># count the number of lines in the file</span>
wc <span class="nt">-l</span> T36R59_I93_S27_L006_R1_sub12M.fastq <span class="c"># (this takes a moment because the file is large)</span>

<span class="c"># print lines with AAAAA in them</span>
<span class="nb">grep</span> <span class="s1">'AAAAA'</span> T36R59_I93_S27_L006_R1_sub12M.fastq <span class="c"># ctrl+c to exit</span>

<span class="c"># count lines with AAAAA in them</span>
<span class="nb">grep</span> <span class="nt">-c</span> <span class="s1">'AAAAA'</span> T36R59_I93_S27_L006_R1_sub12M.fastq 

<span class="c"># save lines with AAAAA in them as a separate file</span>
<span class="nb">grep</span> <span class="s1">'AAAAA'</span> T36R59_I93_S27_L006_R1_sub12M.fastq <span class="o">&gt;</span> AAAAA <span class="c"># no file extensions necessary!!</span>

<span class="c"># add lines with TTTTT to the AAAAA file: '&gt;' writes or overwrites file; '&gt;&gt;' writes or appends to file</span>
<span class="nb">grep</span> <span class="s1">'TTTTT'</span> T36R59_I93_S27_L006_R1_sub12M.fastq <span class="o">&gt;&gt;</span> AAAAA 

<span class="c"># print lines with aaaaa in them</span>
<span class="nb">grep</span> <span class="s1">'aaaaa'</span> T36R59_I93_S27_L006_R1_sub12M.fastq 
<span class="c"># why doesn't this produce any output?</span>

<span class="c"># count number of uniq sequences in file with pattern 'AGAT'</span>
<span class="nb">grep</span> <span class="s1">'AGAT'</span> T36R59_I93_S27_L006_R1_sub12M.fastq | sort | uniq | wc <span class="nt">-l</span>

<span class="c"># print only the second field of the sequence information line</span>
head T36R59_I93_S27_L006_R1_sub12M.fastq | <span class="nb">grep</span> <span class="s1">'@'</span> | awk <span class="nt">-F</span><span class="s1">':'</span> <span class="s1">'{ print $2 }'</span> 
<span class="c"># awk is a very useful program for parsing files; here ':' is the delimiter, $2 is the column location</span>
</code></pre></div></div>

<p><strong>Challenge</strong></p>
<details> 
  <summary>How would you count the number of reads in your file? </summary>
   There are lots of answers for this one. One example: in the fastq format, the character '@' occurs once per sequence, so we can just count how many lines contain '@'.<br /> 
   <code>grep -c '@' T36R59_I93_S27_L006_R1_sub12M.fastq</code>
</details>

<h1 id="assembling-2brad-data-using-ipyrad">Assembling 2bRAD data using iPyrad</h1>

<p><img src="https://github.com/rdtarvin/RADseq_Quito_2017/blob/master/images/basic-assembly-steps.png?raw=true" alt="" /><br /></p>

<p>In this workflow, we will use the <a href="https://github.com/z0on/2bRAD_denovo">2bRAD native pipeline</a> for filtering and trimming, 
<a href="http://hannonlab.cshl.edu/fastx_toolkit/">fastx-toolkit</a> for quality control, 
and then <a href="http://ipyrad.readthedocs.io/index.html">iPyrad</a> for the rest of the assembly.<br /><br /></p>

<h2 id="step-0-use-fastqc-to-check-read-quality">Step 0. Use fastqc to check read quality.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># fastqc is already installed, but if you wanted to download it, this is how: </span>
<span class="c"># wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.5.zip</span>
<span class="c"># unzip fastqc_v0.11.5.zip</span>

<span class="c"># let's take a look at fastqc options</span>
fastqc <span class="nt">-h</span>
fastqc <span class="k">*</span>.fastq
</code></pre></div></div>

<p>When the program is finished, take a look at what files are in the directory using <code class="highlighter-rouge">ls</code>.
fastqc produces a nice .html file that can be viewed in any browser. 
Since we are trying to get comfortable with the command line, let’s open the file directly.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sensible-browser T36R59_I93_S27_L006_R1_sub12M_fastqc.html
</code></pre></div></div>

<p>Sequencing quality scores, “Q”, run from 20 to 40. In the fastq file, these are seen as ASCII characters. 
The values are log-scaled: 20 = 1/100 errors; 30 = 1/1000 errors. Anything below 20 is garbage and anything between 20 and 30 should be reviewed.
There appear to be errors in the kmer content, but really these are just showing where the barcodes and restriction enzyme sites are. 
Let’s take a look at what 2bRAD reads look like:</p>

<p><img src="https://github.com/rdtarvin/RADseq_Quito_2017/blob/master/images/2bRAD-read.png?raw=true" alt="" /></p>

<p>Ok, we can see that overall our data are of high quality (high Q scores, no weird tile patterns, no adaptor contamination). Time to move on to the assembly!!</p>

<h2 id="step-1-demultiplex-by-barcode-in-2brad-native-pipeline">Step 1. Demultiplex by barcode in <strong>2bRAD native pipeline</strong></h2>

<p>Copy scripts to your virtual machine Applications folder via git</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> ~/Applications
git clone https://github.com/z0on/2bRAD_denovo.git
<span class="nb">ls
cd</span> ../workshop/2bRAD/
</code></pre></div></div>
<p>Git is an important programming tool that allows developers to trace changes made in code. Let’s take a look online.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sensible-browser https://github.com/z0on/2bRAD_denovo
</code></pre></div></div>

<p>Okay, before we start the pipeine let’s move our fastqc files to their own directory</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mkdir 2bRAD_fastqc
mv <span class="k">*</span>fastqc<span class="k">*</span> 2bRAD_fastqc <span class="c"># shows error stating that the folder couldn't be moved into the folder</span>
<span class="nb">ls</span> <span class="c"># check that everything was moved</span>
</code></pre></div></div>

<p>First we need to concatenate data that were run on multiple lanes.</p>
<ul>
  <li>T36R59_I93_S27_L00<strong>6</strong>_R1_001.fastq and T36R59_I93_S27_L00<strong>7</strong>_R1_001.fastq</li>
  <li>pattern “T36R59_I93_S27_L00” is common to both read files; program concatenates into one file with (.+) saved as file name</li>
</ul>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>perl ~/Applications/2bRAD_denovo/ngs_concat.pl fastq <span class="s2">"(.+)_S27_L00"</span> <span class="c"># this takes a few minutes so let's take a quick break while it does its thing.</span>
head T36R59_I93.fq

@K00179:73:HJCTJBBXX:7:1101:28006:1209 1:N:0:TGTTAG
TNGTCCACAAGAGTGTGTCGAGCGTTGTGCCCCCCCGCATCTCTACAGAT
+
A#AAFAFJJJJJJJJJJJJJFJJJJJJFJJJJJJJJJJJFJFJJJJJAJ&lt;
@K00179:73:HJCTJBBXX:7:1101:28026:1209 1:N:0:TGTTAG
CNAACCCGTTGTCACGTCGCAAATAAGTCGGTGCGCTGGCAATCACAGAT
+
A#AFFJJJFJJFFJJJJJJJJFFJJJJJFJFJFJFJJJJJJJJJJJJJJF
@K00179:73:HJCTJBBXX:7:1101:28047:1209 1:N:0:TGTTAG
GNGTCCCAGTGATCCGGAGCAGCGACGTCGCTGCTATCCATAGTGAAGAT
</code></pre></div></div>

<p>Now that our read files are concatenated, we need to separate our reads by barcode. 
Let’s take a look again at the structure of a 2bRAD read.<br /></p>

<p><img src="https://github.com/rdtarvin/RADseq_Quito_2017/blob/master/images/2bRAD-read.png?raw=true" alt="" /></p>

<p>From the 2bRAD native pipeline we will use <code class="highlighter-rouge">trim2bRAD_2barcodes_dedup2.pl</code> script to separate by barcode. 
If your data are not from HiSeq4000 you would use <code class="highlighter-rouge">trim2bRAD_2barcodes_dedup.pl</code>. This takes &lt;10 min</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># adaptor is the last 4 characters of the reads, here 'AGAT'</span>
perl ~/Applications/2bRAD_denovo/trim2bRAD_2barcodes_dedup2.pl <span class="nv">input</span><span class="o">=</span>T36R59_I93.fq <span class="nv">adaptor</span><span class="o">=</span>AGAT <span class="nv">sampleID</span><span class="o">=</span>1
<span class="nb">.</span>
<span class="nb">.</span>
<span class="nb">.</span>
T36R59_ACCA: goods: 295784 <span class="p">;</span> dups: 1178405
T36R59_AGAC: goods: 260575 <span class="p">;</span> dups: 987301
T36R59_AGTG: goods: 325595 <span class="p">;</span> dups: 956574
T36R59_CATC: goods: 432507 <span class="p">;</span> dups: 1732082
T36R59_CTAC: goods: 311897 <span class="p">;</span> dups: 1837423
T36R59_GACT: goods: 322958 <span class="p">;</span> dups: 1877945
T36R59_GCTT: goods: 266101 <span class="p">;</span> dups: 1040723
T36R59_GTGA: goods: 483664 <span class="p">;</span> dups: 1871167
T36R59_GTGT: goods: 328948 <span class="p">;</span> dups: 1298015
T36R59_TCAC: goods: 274208 <span class="p">;</span> dups: 1494727
T36R59_TCAG: goods: 309572 <span class="p">;</span> dups: 1109505
T36R59_TGTC: goods: 310155 <span class="p">;</span> dups: 1800440

T36R59: total goods : 3921964 <span class="p">;</span> total dups : 17184307

</code></pre></div></div>

<p>You can see that there are a ton of duplicate reads in these files. This is OK and means that you have high 
coverage across loci. Don’t confuse these duplicates with PCR duplicates, which may be a source of error (more on this later).</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">ls</span>
</code></pre></div></div>

<p>Now you can see that the demultiplexed and deduplicated files are listed by barcode and have the file extension <strong>.tr0</strong>.</p>

<h2 id="step-2-filter-reads-by-quality-with-a-program-from-the-fastx_toolkit">Step 2. Filter reads by quality with a program from the <strong>fastx_toolkit</strong></h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>fastq_quality_filter <span class="nt">-h</span>

<span class="c"># Use typical filter values, i.e., 90% of bases in a read should have a q-value aboce 20</span>
<span class="c"># This is a for loop, which allows you to execute the same command with arguments across multiple files</span>
<span class="c"># The 'i' variable is a place holder </span>
<span class="k">for </span>i <span class="k">in</span> <span class="k">*</span>.tr0<span class="p">;</span> <span class="k">do </span>fastq_quality_filter <span class="nt">-i</span> <span class="nv">$i</span> <span class="nt">-q</span> 20 <span class="nt">-p</span> 90 <span class="o">&gt;</span> <span class="k">${</span><span class="nv">i</span><span class="k">}</span>.trim<span class="p">;</span> <span class="k">done</span>

<span class="c"># zip the files to save space</span>
<span class="k">for </span>i <span class="k">in</span> <span class="k">*</span>.trim<span class="p">;</span> <span class="k">do </span>gzip <span class="k">${</span><span class="nv">i</span><span class="k">}</span><span class="p">;</span> <span class="k">done</span>
</code></pre></div></div>

<p>Now we have our 2bRAD reads separated by barcode and trimmed (steps 1 &amp; 2)!<br /></p>

<p><strong>TASK</strong>: The files need to be renamed for each species. 
Using the barcode file <a href="https://raw.githubusercontent.com/rdtarvin/RADseq_Quito_2017/master/files/2bRAD-ipyrad_barcodes.txt">here</a> and the command <code class="highlighter-rouge">mv</code>, 
rename all .gz files to have the format <code class="highlighter-rouge">genX_spX-X_R1_.fastq.gz</code>.<br /></p>

<h2 id="steps-34567-complete-pipeline-in-ipyrad">Steps 34567. Complete pipeline in <strong>iPyrad</strong></h2>

<p>iPyrad is relatively easy when it comes to command line programs. Make sure you check out their extensive online documentation <a href="http://ipyrad.readthedocs.io/index.html">here</a>.</p>

<p>We need to update our virtual machines before running iPyrad.</p>

<p>Now we can initiate a new analysis.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ipyrad <span class="nt">-n</span> 2brad-v1

<span class="c"># Let's take a look at the file.</span>
<span class="nb">cat </span>params-2brad-v1.txt
</code></pre></div></div>

<p>Make a few changes to the params file.</p>
<ul>
  <li><code class="highlighter-rouge">[4] [sorted_fastq_path]</code>: add the location of the sorted fastqs; you must end this parameter with *.gz</li>
  <li><code class="highlighter-rouge">[7] [datatype]</code>: change rad to gbs; gbs can take into account revcomp reads</li>
  <li><code class="highlighter-rouge">[8] [restriction_overhang]</code>: don’t worry about this, it will be ignored since we are starting from sorted reads</li>
  <li><code class="highlighter-rouge">[11] [mindepth_statistical]</code>: lower to 5; data have been deduplicated (ipyrad assumes it has not been deduplicated)</li>
  <li><code class="highlighter-rouge">[12] [mindepth_majrule]</code>: lower to 2; data have been deduplicated (ipyrad assumes it has not been deduplicated)</li>
  <li><code class="highlighter-rouge">[17] [filter_min_trim_len]</code>: change to 20; loci are small</li>
  <li><code class="highlighter-rouge">[18] [max_alleles_consens]</code>: keep at 2 for diploid</li>
  <li><code class="highlighter-rouge">[21] [min_samples_locus]</code>: change to 3; lower number means more missing data but more loci recovered</li>
  <li><code class="highlighter-rouge">[27] [output_formats]</code>: add ‘, u’; this will provide an output selecting single SNPs from each locus randomly</li>
</ul>

<p><strong>TASK</strong>: Make these changes in the params file using the atom text editor.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>atom params-2brad-v1.txt
</code></pre></div></div>

<p>Your final params file should look like this:</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nt">-------</span> ipyrad params file <span class="o">(</span>v.0.7.1<span class="o">)</span><span class="nt">--------------------------------------------</span>
2brad-v1		               <span class="c">## [0] [assembly_name]: Assembly name. Used to name output directories for assembly steps</span>
./                             <span class="c">## [1] [project_dir]: Project dir (made in curdir if not present)</span>
                               <span class="c">## [2] [raw_fastq_path]: Location of raw non-demultiplexed fastq files</span>
                               <span class="c">## [3] [barcodes_path]: Location of barcodes file</span>
./<span class="k">*</span><span class="nt">-1</span><span class="k">*</span>.gz                         <span class="c">## [4] [sorted_fastq_path]: Location of demultiplexed/sorted fastq files</span>
denovo                         <span class="c">## [5] [assembly_method]: Assembly method (denovo, reference, denovo+reference, denovo-reference)</span>
                               <span class="c">## [6] [reference_sequence]: Location of reference sequence file</span>
gbs                            <span class="c">## [7] [datatype]: Datatype (see docs): rad, gbs, ddrad, etc.</span>
TGCAG,                         <span class="c">## [8] [restriction_overhang]: Restriction overhang (cut1,) or (cut1, cut2)</span>
5                              <span class="c">## [9] [max_low_qual_bases]: Max low quality base calls (Q&lt;20) in a read</span>
33                             <span class="c">## [10] [phred_Qscore_offset]: phred Q score offset (33 is default and very standard)</span>
5                              <span class="c">## [11] [mindepth_statistical]: Min depth for statistical base calling</span>
2                              <span class="c">## [12] [mindepth_majrule]: Min depth for majority-rule base calling</span>
10000                          <span class="c">## [13] [maxdepth]: Max cluster depth within samples</span>
0.85                           <span class="c">## [14] [clust_threshold]: Clustering threshold for de novo assembly</span>
0                              <span class="c">## [15] [max_barcode_mismatch]: Max number of allowable mismatches in barcodes</span>
0                              <span class="c">## [16] [filter_adapters]: Filter for adapters/primers (1 or 2=stricter)</span>
20                             <span class="c">## [17] [filter_min_trim_len]: Min length of reads after adapter trim</span>
2                              <span class="c">## [18] [max_alleles_consens]: Max alleles per site in consensus sequences</span>
5, 5                           <span class="c">## [19] [max_Ns_consens]: Max N's (uncalled bases) in consensus (R1, R2)</span>
8, 8                           <span class="c">## [20] [max_Hs_consens]: Max Hs (heterozygotes) in consensus (R1, R2)</span>
4                              <span class="c">## [21] [min_samples_locus]: Min # samples per locus for output</span>
20, 20                         <span class="c">## [22] [max_SNPs_locus]: Max # SNPs per locus (R1, R2)</span>
8, 8                           <span class="c">## [23] [max_Indels_locus]: Max # of indels per locus (R1, R2)</span>
0.5                            <span class="c">## [24] [max_shared_Hs_locus]: Max # heterozygous sites per locus (R1, R2)</span>
0, 0, 0, 0                     <span class="c">## [25] [trim_reads]: Trim raw read edges (R1&gt;, &lt;R1, R2&gt;, &lt;R2) (see docs)</span>
0, 0, 0, 0                     <span class="c">## [26] [trim_loci]: Trim locus edges (see docs) (R1&gt;, &lt;R1, R2&gt;, &lt;R2)</span>
p, s, v, u                        <span class="c">## [27] [output_formats]: Output formats (see docs)</span>
                               <span class="c">## [28] [pop_assign_file]: Path to population assignment file</span>

</code></pre></div></div>

<p>Okay, it’s that easy!! To run the pipeline, type the following command:</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ipyrad <span class="nt">-p</span> params-2brad-v1.txt <span class="nt">-s</span> 1234567
</code></pre></div></div>

<p>To check on the results, you can open a new terminal window and type</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ipyrad <span class="nt">-p</span> params-2brad-v1.txt <span class="nt">-r</span>
</code></pre></div></div>

<p>Let’s look at the intermediate and final files created by iPyrad.</p>

<h2 id="ipyrad-step-2-filters-the-reads">iPyrad: Step 2. Filters the reads.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd </span>2brad-v1_edits
<span class="nb">ls
cat </span>s2_rawedit_stats.txt 

             reads_raw  trim_adapter_bp_read1  trim_quality_bp_read1  reads_filtered_by_Ns  reads_filtered_by_minlen  reads_passed_filter
gen1_sp1-1a     476680                      0                      0                     0                         0               476680
gen1_sp1-1b     305489                      0                      0                     0                         0               305489
gen2_sp1-1      270185                      0                      0                     0                         0               270185
gen3_sp1-1      307422                      0                      0                     0                         0               307422
gen3_sp2-1      291732                      0                      0                     0                         0               291732
gen3_sp3-1      324320                      0                      0                     0                         0               324320
</code></pre></div></div>

<p><strong>Challenge</strong></p>
<details> 
  <summary>How many reads were filtered and removed from the analysis? Can you explain why? </summary>
   No reads were removed! This is because `[16] [filter_adapters]` was set to 0.
</details>

<p>What do the files created in this step look like?</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>zcat gen1_sp1-1a.trimmed_R1_.fastq.gz | head

@K00179:73:HJCTJBBXX:7:1101:30878:1261 1:N:0:TGTTAG <span class="nv">bcd</span><span class="o">=</span>GTGA
ACTCCCAGCAAGCGATGCTCGTGCATAGCACAGGCG
+
FFJFJJJJA7FAJF-AJJA77F&lt;77FJ&lt;FJ-&lt;FJJJ
@K00179:73:HJCTJBBXX:7:1101:15828:1279 1:N:0:TGTTAG <span class="nv">bcd</span><span class="o">=</span>GTGA
CGGTGCTAACCACGAGACCACTGCCGGTCTGAACTC
+
FFJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJ
@K00179:73:HJCTJBBXX:7:1101:28940:1279 1:N:0:TGTTAG <span class="nv">bcd</span><span class="o">=</span>GTGA
GAGCCTTACAAGGCACGTGTATCGCCTTGACCCGGT
</code></pre></div></div>

<p>They should be the same as the input files, but now they are in a place with names that iPyrad can recognize and use downstream.</p>

<h2 id="ipyrad-step-3-clusters-reads-within-individuals">iPyrad: Step 3. Clusters reads within individuals.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> ../2brad-v1_clust_0.85/
<span class="nb">ls
cat </span>s3_cluster_stats.txt

            clusters_total  hidepth_min clusters_hidepth avg_depth_total avg_depth_mj avg_depth_stat sd_depth_total sd_depth_mj sd_depth_stat filtered_bad_align
gen1_sp1-1a         150131          2.0            81028            3.16         5.00          16.97           9.70       12.92         27.05                  0
gen1_sp1-1b         107801          2.0            45137            2.82         5.36          16.57           8.76       13.12         25.54                  0
gen2_sp1-1          100326          2.0            50419            2.69         4.36          16.39           7.54       10.36         23.66                  0
gen3_sp1-1           98206          2.0            53835            3.11         4.85          18.21           9.81       13.00         28.95                  0
gen3_sp2-1          111136          2.0            49989            2.62         4.60          17.15           8.18       11.90         26.98                  0
gen3_sp3-1          117753          2.0            60484            2.74         4.39          16.97           8.37       11.43         26.95                  0

</code></pre></div></div>
<p>You can use this file to see the coverage of your sequences. <code class="highlighter-rouge">avg_depth_total</code> shows us that the coverage is between 2x and 3x for the samples,
but that when we remove clusters with coverage less than 4 (the value we set for <code class="highlighter-rouge">[21] [min_samples_locus]</code>), average coverage is 4-5x.
<br /><br /></p>

<h2 id="ipyrad-step-4-jointly-estimates-heterozygosity-and-error-in-the-clusters-from-step-3">iPyrad: Step 4. Jointly estimates heterozygosity and error in the clusters from step 3.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cat </span>s4_joint_estimate.txt

              hetero_est  error_est
gen1_sp1-1a    0.054254   0.034079
gen1_sp1-1b    0.047668   0.035110
gen2_sp1-1     0.051664   0.033068
gen3_sp1-1     0.056408   0.033520
gen3_sp2-1     0.066856   0.033735
gen3_sp3-1     0.053274   0.033962
</code></pre></div></div>

<p>What do the files from this step look like?</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>zcat gen1_sp1-1a.clustS.gz | head <span class="nt">-30</span>

0005cdf7c5669643589bf3c1aed198e0<span class="p">;</span><span class="nv">size</span><span class="o">=</span>2<span class="p">;</span><span class="k">*</span>
TTGGAGCTGTTTGCACCGATGTCGGACCACGTCGCA
3a67e9aa131e2212644c4aa08de728bb<span class="p">;</span><span class="nv">size</span><span class="o">=</span>2<span class="p">;</span>-
TTGGAGCTGTTTGCACCGAAGTCGGACCACATTTCA
//
//
0006a5dc14fa85f33625de53cf1f282c<span class="p">;</span><span class="nv">size</span><span class="o">=</span>2<span class="p">;</span><span class="k">*</span>
AAAAGAATGCACCGAGTTGTCTGCGCTGCTGGATCT
963fc159360acaffb21ec45e3261c5c7<span class="p">;</span><span class="nv">size</span><span class="o">=</span>2<span class="p">;</span>+
AAAAGAATGCACCGATTTGTCTGCGCTGCTGGATCT
d009cdfd6a5daee4f9d6533af22b3ec1<span class="p">;</span><span class="nv">size</span><span class="o">=</span>2<span class="p">;</span>-
AAAAGAATGCACCGAATTGTATGCGCTGCTGGATCT
04886c38d49d997edd3dc2bd040a9289<span class="p">;</span><span class="nv">size</span><span class="o">=</span>1<span class="p">;</span>-
AAAAGAATGCACCGAGTTGTCTGCACTGCTGGATCT
af95f304ca97a521f58ef4e330ae7c78<span class="p">;</span><span class="nv">size</span><span class="o">=</span>1<span class="p">;</span>+
AAAAGAATGCACCGAGTTGTCTGCGCTGCTAGATCT
9ff8b5a8d90afe70c722a4840bae1bef<span class="p">;</span><span class="nv">size</span><span class="o">=</span>1<span class="p">;</span>+
AGAAGAATGCACCGATTTGTCTGCGCTGCTGGATCT
37d392c265eb57c543a5ee5340d0d849<span class="p">;</span><span class="nv">size</span><span class="o">=</span>1<span class="p">;</span>-
AAAAGAATGCACCGAATTGTCTGCGCTGCTGGATCT
//
//
000758a24010cc268176cd61ea7270e5<span class="p">;</span><span class="nv">size</span><span class="o">=</span>2<span class="p">;</span><span class="k">*</span>
CTGAGTTGTCACCGAGTCCAATGCAGACTCTGCAAT
02d3f761b88391dc4b6f25fe7d560d57<span class="p">;</span><span class="nv">size</span><span class="o">=</span>1<span class="p">;</span>-
CCAAGTTGTCACCGAGTCCAATGCAGACTCTGCAAT
//
//

</code></pre></div></div>

<p>The first cluster looks like a heterozygote. The second cluster here has a ton of different reads and 
is likely to be paralogous. The third cluster could be either a heterozygote or a homozygote with a sequencing error.
The ‘+’ and ‘-‘ indicate whether the reads are reverse complemented.
<br /><br /></p>

<h2 id="ipyrad-step-5-uses-the-estimates-from-step-4-and-filters-the-clusters">iPyrad: Step 5. Uses the estimates from step 4 and filters the clusters.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> ../2brad-v1_consens/
<span class="nb">ls
cat </span>s5_consens_stats.txt 
            clusters_total filtered_by_depth filtered_by_maxH filtered_by_maxN reads_consens  nsites nhetero heterozygosity
gen1_sp1-1a         150131             69103               20             1445         79563 2874663   43055        0.01498
gen1_sp1-1b         107801             62664               23             1029         44085 1595020   31945        0.02003
gen2_sp1-1          100326             49907               17              795         49607 1790170   24390        0.01362
gen3_sp1-1           98206             44371               10              854         52971 1911651   26765        0.01400
gen3_sp2-1          111136             61147                4              861         49124 1773568   29865        0.01684
gen3_sp3-1          117753             57269               19              877         59588 2150534   29350        0.01365
</code></pre></div></div>

<p>Here we get a more accurate measurement of heterozygosity (i.e., with errors removed). You can note that many 
clusters were removed from having too many Ns, which are sites that could not be called with statistical confidence.
<br /><br /></p>

<h2 id="ipyrad-step-6-cluster-loci-among-individuals">iPyrad: Step 6. Cluster loci among individuals.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cat </span>s6_cluster_stats.txt 
vsearch v2.0.3_linux_x86_64, 3.9GB RAM, 1 cores
/home/user1/Applications/BioBuilds/lib/python2.7/site-packages/bin/vsearch-linux-x86_64 <span class="nt">-cluster_smallmem</span> /home/user1/workshop-test/2bRAD/2brad-v1_consens/2brad-v1_catshuf.tmp <span class="nt">-strand</span> both <span class="nt">-query_cov</span> 0.6 <span class="nt">-minsl</span> 0.5 <span class="nt">-id</span> 0.85 <span class="nt">-userout</span> /home/user1/workshop-test/2bRAD/2brad-v1_consens/2brad-v1.utemp <span class="nt">-notmatched</span> /home/user1/workshop-test/2bRAD/2brad-v1_consens/2brad-v1.htemp <span class="nt">-userfields</span> query+target+qstrand <span class="nt">-maxaccepts</span> 1 <span class="nt">-maxrejects</span> 0 <span class="nt">-fasta_width</span> 0 <span class="nt">-threads</span> 0 <span class="nt">-fulldp</span> <span class="nt">-usersort</span> <span class="nt">-log</span> /home/user1/workshop-test/2bRAD/2brad-v1_consens/s6_cluster_stats.txt 
Started  Thu Jul 27 16:23:07 201712095606 nt <span class="k">in </span>334938 seqs, min 32, max 44, avg 36


      Alphabet  nt
    Word width  8
     Word ones  8
        Spaced  No
        Hashed  No
         Coded  No
       Stepped  No
         Slots  65536 <span class="o">(</span>65.5k<span class="o">)</span>
       DBAccel  100%

Clusters: 230503 Size min 1, max 119, avg 1.5
Singletons: 165189, 49.3% of seqs, 71.7% of clusters


Finished Thu Jul 27 16:25:18 2017
Elapsed <span class="nb">time </span>02:11
Max memory 172.5MB
</code></pre></div></div>

<p>Of all this output, probably the only thing we are interested in is the number of clusters and singletons (i.e., clusters with only one read).
There are a number of other files produced during this step as part of the clustering process:</p>
<ul>
  <li>*.consens.gz are consensus reads for each individual</li>
  <li>2brad-v1_catclust.gz contains the clusters created across individuals.</li>
  <li>*.catg are arrays used to estimate the clusters</li>
  <li>2brad-v1.utemp.sort is a table used to sort the reads
<br /><br /></li>
</ul>

<h2 id="ipyrad-step-7-filters-the-data-and-creates-all-the-final-output-files">iPyrad: Step 7. Filters the data and creates all the final output files.</h2>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> ../2brad-v1_outfiles/
<span class="nb">ls
</span>head <span class="nt">-30</span> 2brad-v1.loci 
gen1_sp1-1a     TACATGAGCACTGAGCTCGGTGRCTGGGTAA
gen1_sp1-1b     TACATGAGCACTGAGCTCGGTGRCTGGGTAA
gen3_sp1-1      TACATGAGCACTGAGCTCGGCG-CTGTGTGC
gen3_sp2-1      TACATGAGCACTGAGCTCGGTAACTGGCTAG
//                                  <span class="nt">--</span><span class="k">*</span>   <span class="nt">--</span> <span class="nt">--</span>|11|
gen1_sp1-1a     TGCTCTTGCAGTGGGATCGGCTGGTGACTTG
gen3_sp1-1      AGCTCTTGCAGTGGGATCGGCTGGTGACTTG
gen3_sp2-1      AGCTTTTGCAGTGGGATCGGCTGGTGACTTG
//              -   -                          |36|
gen1_sp1-1a     GGGACACGCAAGAGTCTCGTYTGGTGGTTAC
gen3_sp1-1      NGGACANGCAAGAGTCTCGTNTGGTGGTTAC
gen3_sp2-1      GGGACANGCAAGAGNCTCGTNTGGTGGTTAC
gen3_sp3-1      GGGACACGCAAGAGTCTCGTCTGGTGGTTAC
//                                  -          |92|
gen1_sp1-1a     CCTCAATGCAGCAAATTCGGCYTGTWCCTTT
gen1_sp1-1b     CCTCAATGCAGMAAAGTCGGCCTGTACCTTT
gen3_sp2-1      CCTCAATGCAGCAAAGTCGGYCTGTACMTTT
//                         -   -    <span class="nt">--</span>   - -   |109|
gen1_sp1-1a     CATCCAAGCACCTGAMTCGRTGGGTYGAAAG
gen1_sp1-1b     CATCCAAGCACCTGACTCGGTGGK-CGAAAA
gen2_sp1-1      CATCCAAGCAACCGCATCGGTGGGTCGAAAG
//                        - - -<span class="k">*</span>   -   - -    -|185|
gen1_sp1-1a     CTGCACGTCGACAGCGCTGCAGCCAATAACGG
gen1_sp1-1b     CTG-ACGTCGACAGCGCTGCAGCCAATAACGG
gen2_sp1-1      CTG-ACGTCGACAGCGCTGCAGCCATTAACGG
gen3_sp3-1      CTGCACGTCGACAGCGCTGCAGCCAATCCGTG

</code></pre></div></div>
<p>The <code class="highlighter-rouge">.loci</code> format is unique to iPyrad and is a nice visualization of the final loci.
<code class="highlighter-rouge">-</code> denotes variable sites; <code class="highlighter-rouge">*</code> denotes phylogenetically informative sites.</p>

<p>Let’s look at another output file.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># complete SNP matrix</span>
less <span class="nt">-S</span> 2brad-v1.snps.phy <span class="c"># use -S to prevent line wrapping, scroll to the right</span>
<span class="c"># press 'q' to leave the file</span>

<span class="c"># SNP matrix with only one SNP per locus</span>
less <span class="nt">-S</span> 2brad-v1.u.snps.phy <span class="c"># notice the difference in the number of characters in the file</span>

</code></pre></div></div>

<p>The full results file, <code class="highlighter-rouge">2brad-v1_stats.txt</code>, has more information on the assembly.</p>

<p><br /><br /><br /></p>
<h2 id="downstream-phylogenetic-analyses">Downstream phylogenetic analyses</h2>

<h3 id="phylogenetics-with-ipyrad-is-easy">Phylogenetics with iPyrad is easy!!</h3>

<p>Estimate a <strong>RAxML</strong> tree with concatenated loci</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda install <span class="nt">-p</span> <span class="s2">"</span><span class="nv">$HOME</span><span class="s2">/Applications/BioBuilds"</span> <span class="nt">-y</span> raxml <span class="nt">-c</span> bioconda
conda install <span class="nt">-p</span> <span class="s2">"</span><span class="nv">$HOME</span><span class="s2">/Applications/BioBuilds"</span> <span class="nt">-y</span> toytree <span class="nt">-c</span> eaton-lab
conda install <span class="nt">-p</span> <span class="s2">"</span><span class="nv">$HOME</span><span class="s2">/Applications/BioBuilds"</span> <span class="nt">-y</span> <span class="nt">-c</span> etetoolkit ete3 ete3_external_apps
conda install <span class="nt">-p</span> <span class="s2">"</span><span class="nv">$HOME</span><span class="s2">/Applications/BioBuilds"</span> <span class="nt">-y</span> <span class="nt">-c</span> anaconda biopython
conda install <span class="nt">-p</span> <span class="s2">"</span><span class="nv">$HOME</span><span class="s2">/Applications/BioBuilds"</span> <span class="nt">-y</span> <span class="nt">-c</span> conda-forge matplotlib
python <span class="c"># start python</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">ipyrad.analysis</span> <span class="k">as</span> <span class="n">ipa</span> 
<span class="kn">import</span> <span class="nn">toytree</span>
<span class="n">rax</span> <span class="o">=</span> <span class="n">ipa</span><span class="o">.</span><span class="n">raxml</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s">'2brad-v1.phy'</span><span class="p">,</span><span class="n">name</span><span class="o">=</span><span class="s">'2brad-v1'</span><span class="p">,</span><span class="n">workdir</span><span class="o">=</span><span class="s">'analysis-raxml'</span><span class="p">)</span>
<span class="k">print</span> <span class="n">rax</span><span class="o">.</span><span class="n">command</span>
<span class="n">rax</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">force</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>
<p>To leave the python environment (<code class="highlighter-rouge">&gt;&gt;&gt;</code>) at any point, type <code class="highlighter-rouge">quit()</code>.
Print a tree in python. Type python to enter the python environment, then type</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Phylo</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">'RAxML_bestTree.2brad-v1'</span><span class="p">,</span> <span class="s">'newick'</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">root_at_midpoint</span><span class="p">()</span>
<span class="n">Phylo</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
</code></pre></div></div>

<p>Estimate a quartets-based tree in <strong>tetrad</strong>, an iPyrad version of <a href="http://evomics.org/learning/phylogenetics/svdquartets/">SVDquartets</a></p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tetrad <span class="nt">-s</span> 2brad-v1.snps.phy <span class="nt">-l</span> 2brad-v1.snps.map <span class="nt">-m</span> all <span class="nt">-n</span> 2brad-v1
</code></pre></div></div>

<p>Start python to view the tree by typing <code class="highlighter-rouge">python</code>, then type:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Phylo</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">'analysis-tetrad/2brad-v1.tree'</span><span class="p">,</span> <span class="s">'newick'</span><span class="p">)</span>
<span class="n">Phylo</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span> <span class="c"># is ultrametric</span>
<span class="n">quit</span><span class="p">()</span>
</code></pre></div></div>

<p>Estimate a tree based on a SNP matrix, using one SNP from each locus, in <a href="https://github.com/amkozlov/raxml-ng">RAxML-ng</a>.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cd</span> ~/Applications
wget https://github.com/amkozlov/raxml-ng/releases/download/0.4.0/raxml-ng_v0.4.0b_linux_x86_64.zip
unzip raxml-ng_v0.4.0b_linux_x86_64.zip
raxml-ng <span class="nt">-h</span>
rm raxml-ng_v0.4.0b_linux_x86_64.zip
<span class="nb">.</span>
<span class="nb">.</span>
<span class="nb">.</span>
<span class="nb">cd</span> ../workshop/2brad/2brad-v1_outfiles/
<span class="nv">$HOME</span>/Applications/raxml-ng <span class="nt">--msa</span> 2brad-v1.u.snps.phy <span class="nt">--model</span> GTR+G+ASC_LEWIS <span class="nt">--search</span>
</code></pre></div></div>
<p>Note that there are three options for ascertainment bias correction.</p>

<p>Start python to view the tree by typing <code class="highlighter-rouge">python</code>, then type:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Phylo</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">'2brad-v1.u.snps.phy.raxml.bestTree'</span><span class="p">,</span> <span class="s">'newick'</span><span class="p">)</span>
<span class="n">Phylo</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span> <span class="c"># is ultrametric</span>
<span class="n">quit</span><span class="p">()</span>
</code></pre></div></div>

<p>This is the expected topology and estimated divergence timing.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>        ___________________gen1_sp1
       |   
-25my--|          _________gen2_sp1
       |____15my_|
                 |      ___gen3_sp1
                 |_5my_|  
                       |  _gen3_sp2
                       |_|
                         |_gen3_sp3
                         
</code></pre></div></div>

<p>Do we see the correct tree in our results?</p>

<h3 id="further-exercise">Further exercise</h3>

<p>iPyrad offers the option to branch your pipeline with argument -b. This allows you to access results from previous steps and use them in different downstream analyses.
One of the most convenient uses of this function is testing the effect of missing data on your downstream analyses.
So, let’s execute some reiterations of Step 7, the step that generates the final SNP matrices.
<strong>Remember, never edit file names or paths that are created by iPyrad. It needs them to keep track of analyses.</strong></p>

<p>First, check when the [21] missing data parameter is used in the analysis, to be sure that you don’t need to rerun other steps in addition to 7.
This information can be found in the iPyrad documents <a href="http://ipyrad.readthedocs.io/parameters.html#">parameters page</a>. Scroll down to [21], what does it say?</p>

<p><code class="highlighter-rouge">Affected steps = 7. Example entries to params.txt</code><br /></p>

<p>Great, so we can just redo Step 7 with new values for [21]. The basic command to use here is</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ipyrad <span class="nt">-p</span> params-2brad-v1.txt <span class="nt">-b</span> 2brad-v2-6min
<span class="c"># provide new, informative prefix</span>
</code></pre></div></div>

<p>Then you would need to open the new params file (<code class="highlighter-rouge">params-2brad-v2-6min.txt</code>) and manually edit parameter [21] to be 6.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>atom params-2brad-v2-6min.txt
</code></pre></div></div>

<p>If you had more samples, you could automate this process as such.</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for </span>i <span class="k">in </span>4 6 8 10 12<span class="p">;</span> <span class="k">do </span>ipyrad <span class="nt">-p</span> params-2brad-v1.txt <span class="nt">-b</span> 2brad-v2-<span class="k">${</span><span class="nv">i</span><span class="k">}</span>l<span class="p">;</span> <span class="k">do </span>sed <span class="nt">-i</span> s/<span class="s2">".*</span><span class="se">\[</span><span class="s2">21</span><span class="se">\]</span><span class="s2">.*"</span>/<span class="s2">"</span><span class="k">${</span><span class="nv">i</span><span class="k">}$(</span><span class="nb">printf</span> <span class="s1">'\t'</span><span class="k">)$(</span><span class="nb">printf</span> <span class="s1">'\t'</span><span class="k">)$(</span><span class="nb">printf</span> <span class="s1">'\t'</span><span class="k">)$(</span><span class="nb">printf</span> <span class="s1">'\t'</span><span class="k">)</span><span class="s2"> </span><span class="se">\#\#</span><span class="s2"> </span><span class="se">\[</span><span class="s2">21</span><span class="se">\]</span><span class="s2"> </span><span class="se">\[</span><span class="s2">min_samples_locus</span><span class="se">\]</span><span class="s2">: Min </span><span class="se">\#</span><span class="s2"> samples per locus for output"</span>/ params-2brad-v2-<span class="k">${</span><span class="nv">i</span><span class="k">}</span>l.txt<span class="p">;</span>  <span class="k">done</span>
<span class="c"># for i in 4 6 8 10 12; do ipyrad -p params-2brad-v2-${i}l.txt -s 7 -f; done</span>

<span class="c">## use the '-f' option when you have already run that step, to force it to rerun</span>
</code></pre></div></div>
<p><strong>TASK</strong>: Choose one params file that is different from your neighbor and run it so we can compare results as a class.</p>

<p><strong>TASK</strong>: It’s a good idea to run iPyrad using both a range of [21] missing data (as above)
and of [14] clustering threshold. Choose one value lower and one value higher than 0.85 for parameter [21], create two new branches, and run these analyses using the -b option in iPyrad.<br />
<strong>Hint</strong>: First, be sure to check when the clustering threshold was incorporated into the analysis before deciding where to restart the pipeline.</p>

<p>A full run of these analyses can be downloaded <a href="https://www.dropbox.com/s/w9o7diw6h6lib1m/2brad-ipyrad-full-run.zip?dl=1">here</a></p>

<p><a href="https://rdtarvin.github.io/RADseq_Quito_2017/"><button>Home</button></a>    <a href="https://rdtarvin.github.io/RADseq_Quito_2017/main/2017/08/02/afternoon-ddRAD-pyrad.html"><button>Next Lesson</button></a></p>

<h2 id="appendix">Appendix.</h2>

<p>If you get “Segmentation Errors”, exit Virtual Box and close the program. You may need to allocate more memory to the system.</p>

<p>Step 1. Allocate more space to the drive.<br />
Run the following commands through the native OSX Terminal program. <a href="https://www.jeffgeerling.com/blogs/jeff-geerling/resizing-virtualbox-disk-image">Source</a></p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># in OSX (for Windows the commands will be slightly different)</span>
<span class="nb">cd</span> /Users/&lt;your user name&gt;/VirtualBox<span class="se">\ </span>VMs/UT<span class="se">\ </span>Biocomputing/
<span class="c"># Clone the .vmdk image to a .vdi.</span>
vboxmanage clonehd <span class="s1">'UT Biocomputing-disk001.vmdk'</span> <span class="s1">'new_UT Biocomputing-disk001.vdi'</span> <span class="nt">--format</span> vdi
<span class="c"># Resize the new .vdi image (51200 == 50 GB).</span>
vboxmanage modifyhd <span class="s2">"new_UT Biocomputing-disk001.vdi"</span> <span class="nt">--resize</span> 51200 
</code></pre></div></div>

<p>Step 2. Tell the VM how to access that space. <a href="[here](https://www.howtoforge.com/partitioning_with_gparted)">Source</a>. If your cursor gets stuck, press the Command key.</p>

<ol>
  <li>Open the Virtual Box program and go to Settings, then Storage, and click on SATA. Click on the option to add a new hard disk, select the ‘new’ disk that you created and press OK.</li>
  <li>Download <a href="http://gparted.org/download.php">gparted</a></li>
  <li>In Virtual Box, click Settings, then Storage, then click on Controller: IDE and select … Click the box that says ‘Live CD/DVD’</li>
  <li>Now go to Storage and unclick the box that says Hard Disk.</li>
  <li>Start the VM and click enter through the language settings, selecting those that are appropriate for you.</li>
  <li>On the upper right hand corner, make sure the new disk is selected (you’ll see one bar with linux-swap, another called extended, and an open grey box with dashed outline)</li>
  <li>Click ‘New’, allow 1MB to remain unallocated, and press OK. Then press Apply.</li>
  <li>Turn off the VM. Go to Settings, Storage, right click on the GParted disk and click Remove.</li>
  <li>Click on System, click Hard Disk, and make sure it is on the top of the list, using the arrows. Make sure there is a Hard Drive assigned to Port 1.</li>
  <li>Start the VM again!</li>
</ol>



      <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="https://github.com/rdtarvin/IBS2019_Genomics-of-Biodiversity">IBS2019_Genomics-of-Biodiversity</a> is maintained by <a href="https://github.com/rdtarvin">rdtarvin</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer>
    </section>

    
  </body>
</html>
